src/sage/rings/finite_rings/integer_mod_ring.py

```diff
@@ -1547,6 +1551,418 @@ def random_element(self, bound=None):
+        time.  While we can no longer use Newton's method to solve for a lift,
+        the Taylor series it is based on still yields constraints on the roots
+        modulo `p^{k+1}`:  If `r_k` is a root of `f` modulo `p^k`, then either
+        every lift of `r_k` to `\Zmod{p^{k + 1}}` is a root of `f`
+        modulo `p^{k+1}` or none of them are.  Consequently we may find roots
+        modulo `p^e` by lifting one power at a time.
+
+        When `f'(r) \equiv 0 \pmod{p}`, an alternative approach is to change
+        variables, factor out the root, and then factor out powers of `p`.
+        This has the advantage that it will eventually reach a situation where
+        the lift converges quadratically, but it is not presently implemented.
+        A different form of Hensel's lemma applies once we are close enough to
+        a `p`-adic root (see [Gou2020]_, problem 120), but it seems delicate
+        to use it directly to find all roots modulo `p^e` (consider our
+        earlier example of `f(x) = 2x`), so we do not presently attempt to
+        apply Hensel's lemma in this way.
+
+        EXAMPLES::
+
+            sage: R.<x> = Zmod(41)[]
+            sage: (x^3 + x).roots()
+            [(0, 1), (32, 1), (9, 1)]
+            sage: (x^3 + x).roots(multiplicities=False)
+            [0, 32, 9]
+            sage: (x^6 + x^5 + 9*x^4 + 20*x^3 + 3*x^2 + 18*x + 7).roots()
+            [(19, 1), (20, 2), (21, 3)]
+            sage: (x^6 + x^5 + 9*x^4 + 20*x^3 + 3*x^2 + 18*x + 7).roots(multiplicities=False)
+            [19, 20, 21]
+
+        We can find roots without multiplicities over a ring whose modulus is
+        a prime power, even a big power:
+
+            sage: R.<x> = Zmod(7^3)[]
+            sage: (x^2 + x + 1).roots(multiplicities=False)
+            [18, 324]
+            sage: R.<x> = Zmod(2^50)[]
+            sage: (x + 1).roots(multiplicities=False)
+            [1125899906842623]
+
+        We can also find roots without multiplicities over a ring whose modulus
+        is a product of primes or prime powers:
+
+            sage: R.<x> = Zmod(60)[]
+            sage: (x^2 - 1).roots(multiplicities=False)
+            [29, 41, 49, 1, 59, 11, 19, 31]
+
+        We may also ask for roots modulo a quotient of the ring over which the
+        polynomial is defined:
+
+            sage: R.<x> = Zmod(120)[]
+            sage: (x^2 - 1).roots(multiplicities=False)
+            [89, 41, 49, 1, 29, 101, 109, 61, 59, 11, 19, 91, 119, 71, 79, 31]
+            sage: (x^2 - 1).roots(Zmod(60), multiplicities=False)
+            [29, 41, 49, 1, 59, 11, 19, 31]
+
+        TESTS::
+
+            sage: R.<x> = Zmod(2)[]
+            sage: x.roots()
+            [(0, 1)]
+
+        Test polynomials with content:
+
+            sage: R.<x> = Zmod(4)[]
+            sage: (2*x).roots(multiplicities=False)
+            [0, 2]
+
+            sage: R.<x> = Zmod(6)[]
+            sage: (3*x).roots(multiplicities=False)
+            [0, 4, 2]
+
```

Comment:
```suggestion

        Test polynomial with many roots:

                    sage: R.<x> = Zmod(6)[]
                    sage: f = x * (x - 1) * (x - 2) * (x - 3) * (x - 4) * (x - 5)
                    sage: len(f.roots(multiplicities=False))
                    6

```

Reply:
This works as-is. The roots are deduplicated in the base case, when we look for roots over finite fields. They can't reappear during the CRT because the CRT is an isomorphism.